articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+
+[To discuss: Should there be a callback for when configuration is changed, e.g. altered parameters, new topics, etc.?]
+
+While in this state, the node will not receive any execution time to read topics, perform processing of data, respond to functional service requests, etc.
+
+In the inactive state, any data that arrives on topics will not be read. Whether or not data is queued will depend on the QoS parameters of the topics involved. [In other words, to queue data or not is up to each topic.]
+
+Any service requests to a node in the inactive state will not be answered (to the caller, they will fail immediately).
+
+Entering the `inactive` state from the `created` state requires running the node's `onInitialised` method. The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the node's `onDeactivated` method. See the `active` state description for more information.
+
+Entering the `inactive` state from the `fatal error` state requires running the node's `onReset` method. See the `fatal error` state description for more information.
+
+If an error occurs while in this state that cannot be handled by the node, then the transition to the `fatal error` state is taken.
+
+The node may also exit this state by being destroyed, in which case the `destroy` transition is taken.
+
+Implementation note: Causing the transition to the destroyed state can often be mapped to deletion of the node instance.
+
+### Active state
+
+This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
+
+Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
+
+To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
+
+If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
+
+### Fatal Error state
+
+Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
```

Comment:
> An "error" is by definition something the node should not recover for.

By what definition? I think we're mixing the terms, because I don't know of any definition of error which excludes the idea that the actor should try to recover from it. To me that would be a "fatal error" or a "fault" or an "error which cannot be mitigated".  If the node has no error recovery logic in it then you're saying that all error recovery logic must either be in the supervision code (which breaks the encapsulation of the node) or the resolution for all errors must be either retry or die or ignore (basically the options provided by roslaunch).

So, @doudou, do you not think that it makes sense for non-trivial error recovery behavior to be captured in a node which actually knows something about the context in which the error occurred?

If no, then does that mean you now need the node and special logic which exists outside the node in order to use it effectively (in the case of non-trivial error recovery)? What benefit does separating the error handling logic from the normal operation logic provide?

If yes, then my other point was that if nodes wish to do some error recovery before giving up then we might need to have a life cycle state for them to do it in. This has two benefits, one is that it exposes the fact that error recovery is happening through the life cycle, rather than having a delayed transition or interrupted state activity while error recovery is on going. The other benefit would be that this error recovery state could have different constraints, like allowing for creation of publishers and subscribers, or allowing for memory allocation, or changes to configuration, all of which might be prohibited in other states or transitions. The first point could be accomplished using diagnostics or console logging or something like this, but as long as you have events for life cycle it might be a convenient way to notify the system. Everything in this paragraph is sort of speculative, but I think it's worth thinking about despite the increased complexity to the life cycle design.